### PR TITLE
Fix #4703: move IfInner containers to the end

### LINKED ISSUE #4703: Inlined function with update stage won't vectorize when parent function is scheduled with compute_with
In the following example:

```
    const int f_size = 128;
    const int g_size = 256;
    Buffer<int> f_im(f_size, f_size, 5), g_im(g_size, g_size);

    Var x("x"), y("y"), c("c"), xi("xi"), yi("yi"), yii("yii"), yo("yo");
    Func f("f"), g("g"), h("h"), input("input");

    input(x, y) = x;
    f(x, y, c) = c * input(x, y);
    h(x, y, c) = f(x, y, c);

    Func inl("inl");
    inl(x, y) = f(x / 2, y / 2, 0);
    inl(x, y) += f(x / 2, y / 2, 2);
    g(x, y) = inl(x, y);

    g
        .split(y, yo, y, 32 * 2, TailStrategy::RoundUp)
        .split(y, y, yi, 2, TailStrategy::RoundUp)
        .vectorize(x, 4, TailStrategy::GuardWithIf)
        .compute_with(h, y, LoopAlignStrategy::AlignEnd)
        //.compute_root()
        ;

    h
        .reorder(x, c, y)
        .split(y, yo, y, 32, TailStrategy::RoundUp)
        .split(y, y, yi, 1, TailStrategy::RoundUp)
        .vectorize(x, 4, TailStrategy::GuardWithIf)
        .compute_root();

    g.bound(y, 0, g_size);
    h.bound(y, 0, f_size).bound(c, 0, 5);

    Pipeline p({h, g});
    p.compile_jit();
```

Here function `inl` is inlined into `g`, which is scheduled to be computed with `f`.  `g` is vectorized across x, but in the final IR `inl` serialized and compute in the loop: 

```
      for (g.s0.x.x, 0, t266) {
       allocate inl[int32 * 4]
       let inl.s0.x.loop_extent.s = (g.extent.0 - (g.s0.x.x*4))
       produce inl {
        let t279 = max(min(inl.s0.x.loop_extent.s, 4), 0)
        let t280 = (g.s0.x.x*2)
        for (g.s0.x.v0, 0, t279) {
         inl[g.s0.x.v0] = 0
         inl[g.s0.x.v0] = (inl[g.s0.x.v0] + ((((g.min.0 + g.s0.x.v0)/2) + t280)*2))
        }
       }
       consume inl {
        g[ramp(((g.s0.x.x*4) + ((g.stride.1*t278) + t264)), 1, 4)] = inl[ramp(0, 1, 4)]
       }
       free inl
      }

```
However, if we schedule `g` as compute_root, `inl` will be properly vectorized:

```
    for (g.s0.x.x, 0, t50) {
      allocate inl[int32 * 4]
      produce inl {
       inl[ramp(0, 1, 4)] = x4(0)
       inl[ramp(0, 1, 4)] = interleave_vectors((ramp((((g.s0.x.x*2) + t46)*2), 2, 2) + inl[ramp(0, 2, 2)]), (ramp((((g.s0.x.x*2) + t47)*2), 2, 2) + inl[ramp(1, 2, 2)]))
      }
      consume inl {
       g[ramp(((g.s0.x.x*4) + ((g.stride.1*t53) + t48)), 1, 4)] = inl[ramp(0, 1, 4)]
      }
      free inl
     }
```

### PR DESCRIPTION
**Pull request by vksnk**
As described in #4703, in some cases when function is scheduled with compute_with, inner loop won't be properly vectorized . It seems that this is happening because bounds for the rebased variable in the said loop are not tight enough. If you look at the IR right after initial scheduling:

```
        vectorized (g.s0.x.v0, g.s0.x.v0.loop_min, g.s0.x.v0.loop_extent) {
         if ((uint1)likely((g.s0.y.fused.yo >= g.s0.y.yo.loop_min))) {
          if ((uint1)likely((g.s0.y.fused.yo <= g.s0.y.yo.loop_max))) {
           if ((uint1)likely(((g.s0.y.y.fused.y + (g.s0.y.y.y.loop_max - h.s0.y.y.y.loop_max)) >= g.s0.y.y.y.loop_min))) {
            if ((uint1)likely(((g.s0.y.y.fused.y + (g.s0.y.y.y.loop_max - h.s0.y.y.y.loop_max)) <= g.s0.y.y.y.loop_max))) {
             let g.s0.x.rebased = ((g.s0.x.x*4) + g.s0.x.v0)
             if ((uint1)likely((g.s0.x.rebased < g.s0.x.loop_extent))) {
              let g.s0.x = (g.s0.x.v0.base + g.s0.x.v0)

```

g.s0.x.rebased ends up in between two ifs and I think this makes it harder for the bounds inference to reason about it (there is actually a [comment](https://github.com/halide/Halide/blob/master/src/ScheduleFunctions.cpp#L163), which says something like this and specifically mentions rebased vars; well, I can convince myself that it says exactly the opposite, but I like it other way, because it supports my theory).

This PR attempts to fix the issue by moving `IfInner` containers to the end and giving higher 'priority' to the `If`containers during sorting. In other words, all `If` containers will go first in the outward order with `IfInner` going after and let vars still being the most inward (assuming that there are no ifs that depend on them). That being said, I don't think the order of ifs matters on its own, instead depended if, basically, serves as a guard for var it depends on against `IfInner`, which can't go outside of the loop (regular ifs can) and guaranteed to stay there.

After the change IR will look like:

```
        vectorized (g.s0.x.v0, g.s0.x.v0.loop_min, g.s0.x.v0.loop_extent) {
         let g.s0.x.rebased = ((g.s0.x.x*4) + g.s0.x.v0)
         if ((uint1)likely((g.s0.x.rebased < g.s0.x.loop_extent))) {
          if ((uint1)likely((g.s0.y.fused.yo >= g.s0.y.yo.loop_min))) {
           if ((uint1)likely((g.s0.y.fused.yo <= g.s0.y.yo.loop_max))) {
            if ((uint1)likely(((g.s0.y.y.fused.y + (g.s0.y.y.y.loop_max - h.s0.y.y.y.loop_max)) >= g.s0.y.y.y.loop_min))) {
             if ((uint1)likely(((g.s0.y.y.fused.y + (g.s0.y.y.y.loop_max - h.s0.y.y.y.loop_max)) <= g.s0.y.y.y.loop_max))) {
              let g.s0.x = (g.s0.x.v0.base + g.s0.x.v0)

```

This definitely needs a test (so far I just basically eyeballed generated IR), but wanted to send earlier to check if this makes sense at all .

### CODE REVIEW

File: src/ScheduleFunctions.cpp [hunk 1]
```diff
@@ -206,14 +206,19 @@ Stmt build_loop_nest(
...
     }
 
-    // Sort the predicate guards for the fused loops so they are as far outwards
-    // as possible. IfInnner should not be reordered to outside of a for loop.
-    for (int i = (int)nest.size() - n_predicates_inner - n_predicates; i < (int)nest.size() - n_predicates; i++) {
+    // Sort the ifs so they are as far outwards as possible.
+    // BoxesTouched trims the domain of a variable within a scope of if-then-else
+    // based on the likely condition. However, it doesn't do it transitively; it
+    // doesn't trim the domains of other variables that directly/indirectly
+    // depend on the original variable in the likely condition outside the scope
```
Comment (abadams) [inline src/ScheduleFunctions.cpp]: I believe it *does* do it transitively. There's a lot of logic to handle trimming bounds of dependent vars. It's certainly harder for it though.

Comment (abadams) [reply src/ScheduleFunctions.cpp]: Oh you just moved this comment. It must have been stale.

**Comment by abadams** [comment]
In your example the extra ifs don't seem relevant to the loop over x and I don't see how they could affect things. Can you paste the full IR just before vectorization of the failure case with and without this change?

**Comment by vksnk** [comment]
> In your example the extra ifs don't seem relevant to the loop over x and I don't see how they could affect things. Can you paste the full IR just before vectorization of the failure case with and without this change?

Thank you for looking into this! Here is the IRs right before vectorization pass:

before:

```
assert(((uint64)reinterpret(((halide_buffer_t *))h.buffer) != (uint64)0), halide_error_buffer_argument_is_null("h"))
assert(((uint64)reinterpret(((halide_buffer_t *))g.buffer) != (uint64)0), halide_error_buffer_argument_is_null("g"))
let g = ((void *))_halide_buffer_get_host(((halide_buffer_t *))g.buffer)
let g.min.0 = _halide_buffer_get_min(((halide_buffer_t *))g.buffer, 0)
let g.extent.0 = _halide_buffer_get_extent(((halide_buffer_t *))g.buffer, 0)
let g.stride.0 = _halide_buffer_get_stride(((halide_buffer_t *))g.buffer, 0)
let g.min.1 = _halide_buffer_get_min(((halide_buffer_t *))g.buffer, 1)
let g.stride.1 = _halide_buffer_get_stride(((halide_buffer_t *))g.buffer, 1)
let h = ((void *))_halide_buffer_get_host(((halide_buffer_t *))h.buffer)
let h.min.0 = _halide_buffer_get_min(((halide_buffer_t *))h.buffer, 0)
let h.extent.0 = _halide_buffer_get_extent(((halide_buffer_t *))h.buffer, 0)
let h.stride.0 = _halide_buffer_get_stride(((halide_buffer_t *))h.buffer, 0)
let h.min.1 = _halide_buffer_get_min(((halide_buffer_t *))h.buffer, 1)
let h.stride.1 = _halide_buffer_get_stride(((halide_buffer_t *))h.buffer, 1)
let h.min.2 = _halide_buffer_get_min(((halide_buffer_t *))h.buffer, 2)
let h.stride.2 = _halide_buffer_get_stride(((halide_buffer_t *))h.buffer, 2)
if ((uint1)_halide_buffer_is_bounds_query(((halide_buffer_t *))g.buffer)) {
 ((halide_buffer_t *))_halide_buffer_init(((halide_buffer_t *))g.buffer, ((halide_dimension_t *))_halide_buffer_get_shape(((halide_buffer_t *))g.buffer), ((void *))reinterpret((uint64)0), (uint64)0, ((halide_device_interface_t *))reinterpret((uint64)0), 0, 32, 2, ((halide_dimension_t *))make_struct(g.min.0, g.extent.0, 1, 0, 0, 256, g.extent.0, 0), (uint64)0)
}
if ((uint1)_halide_buffer_is_bounds_query(((halide_buffer_t *))h.buffer)) {
 ((halide_buffer_t *))_halide_buffer_init(((halide_buffer_t *))h.buffer, ((halide_dimension_t *))_halide_buffer_get_shape(((halide_buffer_t *))h.buffer), ((void *))reinterpret((uint64)0), (uint64)0, ((halide_device_interface_t *))reinterpret((uint64)0), 0, 32, 3, ((halide_dimension_t *))make_struct(h.min.0, h.extent.0, 1, 0, 0, 128, h.extent.0, 0, 0, 5, (h.extent.0*128), 0), (uint64)0)
}
if (!((uint1)_halide_buffer_is_bounds_query(((halide_buffer_t *))g.buffer) || (uint1)_halide_buffer_is_bounds_query(((halide_buffer_t *))h.buffer))) {
 assert((g.stride.0 == 1), 0)
 assert((h.stride.0 == 1), 0)
 produce g {
  produce h {
   for (h.s0.y.fused.yo, 0, 4) {
    for (h.s0.y.y.fused.y, 0, 32) {
     for (h.s0.c, 0, 5) {
      for (h.s0.x.x, 0, ((h.extent.0 + 3)/4)) {
       vectorized (h.s0.x.v1, 0, 4) {
        if ((uint1)likely((uint1)1)) {
         if ((uint1)likely((((h.s0.x.x*4) + h.s0.x.v1) < h.extent.0))) {
          h[(((h.s0.c*h.stride.2) + ((((h.s0.y.fused.yo*32) + h.s0.y.y.fused.y)*h.stride.1) + (((h.s0.x.x*4) + h.s0.x.v1) + h.min.0))) - ((h.min.2*h.stride.2) + ((h.min.1*h.stride.1) + h.min.0)))] = ((((h.s0.x.x*4) + h.s0.x.v1) + h.min.0)*h.s0.c)
         }
        }
       }
      }
     }
     for (g.s0.y.y.yi, 0, 2) {
      for (g.s0.x.x, 0, ((g.extent.0 + 3)/4)) {
       vectorized (g.s0.x.v0, 0, 4) {
        let g.s0.y.min_3 = ((h.s0.y.fused.yo*64) + ((h.s0.y.y.fused.y*2) + g.s0.y.y.yi))
        if ((uint1)likely((uint1)1)) {
         if ((uint1)likely((((g.s0.x.x*4) + g.s0.x.v0) < g.extent.0))) {
          allocate inl[int32 * 1 * 1]
          let inl.s0.x.loop_extent.s = (g.extent.0 - (g.s0.x.x*4))
          produce inl {
           if ((g.s0.x.v0 < inl.s0.x.loop_extent.s)) {
            inl[0] = 0
            inl[0] = (inl[0] + ((((g.min.0 + g.s0.x.v0)/2) + (g.s0.x.x*2))*2))
           }
          }
          consume inl {
           g[(((g.s0.y.min_3*g.stride.1) + (((g.s0.x.x*4) + g.s0.x.v0) + g.min.0)) - ((g.min.1*g.stride.1) + g.min.0))] = inl[0]
          }
         }
        }
       }
      }
     }
    }
   }
  }
 }
}

```
after the change:

```
assert(((uint64)reinterpret(((halide_buffer_t *))h.buffer) != (uint64)0), halide_error_buffer_argument_is_null("h"))
assert(((uint64)reinterpret(((halide_buffer_t *))g.buffer) != (uint64)0), halide_error_buffer_argument_is_null("g"))
let g = ((void *))_halide_buffer_get_host(((halide_buffer_t *))g.buffer)
let g.min.0 = _halide_buffer_get_min(((halide_buffer_t *))g.buffer, 0)
let g.extent.0 = _halide_buffer_get_extent(((halide_buffer_t *))g.buffer, 0)
let g.stride.0 = _halide_buffer_get_stride(((halide_buffer_t *))g.buffer, 0)
let g.min.1 = _halide_buffer_get_min(((halide_buffer_t *))g.buffer, 1)
let g.stride.1 = _halide_buffer_get_stride(((halide_buffer_t *))g.buffer, 1)
let h = ((void *))_halide_buffer_get_host(((halide_buffer_t *))h.buffer)
let h.min.0 = _halide_buffer_get_min(((halide_buffer_t *))h.buffer, 0)
let h.extent.0 = _halide_buffer_get_extent(((halide_buffer_t *))h.buffer, 0)
let h.stride.0 = _halide_buffer_get_stride(((halide_buffer_t *))h.buffer, 0)
let h.min.1 = _halide_buffer_get_min(((halide_buffer_t *))h.buffer, 1)
let h.stride.1 = _halide_buffer_get_stride(((halide_buffer_t *))h.buffer, 1)
let h.min.2 = _halide_buffer_get_min(((halide_buffer_t *))h.buffer, 2)
let h.stride.2 = _halide_buffer_get_stride(((halide_buffer_t *))h.buffer, 2)
if ((uint1)_halide_buffer_is_bounds_query(((halide_buffer_t *))g.buffer)) {
 ((halide_buffer_t *))_halide_buffer_init(((halide_buffer_t *))g.buffer, ((halide_dimension_t *))_halide_buffer_get_shape(((halide_buffer_t *))g.buffer), ((void *))reinterpret((uint64)0), (uint64)0, ((halide_device_interface_t *))reinterpret((uint64)0), 0, 32, 2, ((halide_dimension_t *))make_struct(g.min.0, g.extent.0, 1, 0, 0, 256, g.extent.0, 0), (uint64)0)
}
if ((uint1)_halide_buffer_is_bounds_query(((halide_buffer_t *))h.buffer)) {
 ((halide_buffer_t *))_halide_buffer_init(((halide_buffer_t *))h.buffer, ((halide_dimension_t *))_halide_buffer_get_shape(((halide_buffer_t *))h.buffer), ((void *))reinterpret((uint64)0), (uint64)0, ((halide_device_interface_t *))reinterpret((uint64)0), 0, 32, 3, ((halide_dimension_t *))make_struct(h.min.0, h.extent.0, 1, 0, 0, 128, h.extent.0, 0, 0, 5, (h.extent.0*128), 0), (uint64)0)
}
if (!((uint1)_halide_buffer_is_bounds_query(((halide_buffer_t *))g.buffer) || (uint1)_halide_buffer_is_bounds_query(((halide_buffer_t *))h.buffer))) {
 assert((g.stride.0 == 1), 0)
 assert((h.stride.0 == 1), 0)
 produce g {
  produce h {
   for (h.s0.y.fused.yo, 0, 4) {
    for (h.s0.y.y.fused.y, 0, 32) {
     for (h.s0.c, 0, 5) {
      for (h.s0.x.x, 0, ((h.extent.0 + 3)/4)) {
       vectorized (h.s0.x.v1, 0, 4) {
        if ((uint1)likely((((h.s0.x.x*4) + h.s0.x.v1) < h.extent.0))) {
         if ((uint1)likely((uint1)1)) {
          h[(((h.s0.c*h.stride.2) + ((((h.s0.y.fused.yo*32) + h.s0.y.y.fused.y)*h.stride.1) + (((h.s0.x.x*4) + h.s0.x.v1) + h.min.0))) - ((h.min.2*h.stride.2) + ((h.min.1*h.stride.1) + h.min.0)))] = ((((h.s0.x.x*4) + h.s0.x.v1) + h.min.0)*h.s0.c)
         }
        }
       }
      }
     }
     for (g.s0.y.y.yi, 0, 2) {
      for (g.s0.x.x, 0, ((g.extent.0 + 3)/4)) {
       vectorized (g.s0.x.v0, 0, 4) {
        let g.s0.y.min_3 = ((h.s0.y.fused.yo*64) + ((h.s0.y.y.fused.y*2) + g.s0.y.y.yi))
        if ((uint1)likely((((g.s0.x.x*4) + g.s0.x.v0) < g.extent.0))) {
         if ((uint1)likely((uint1)1)) {
          allocate inl[int32 * 1 * 1]
          produce inl {
           inl[0] = 0
           inl[0] = (inl[0] + ((((g.min.0 + g.s0.x.v0)/2) + (g.s0.x.x*2))*2))
          }
          consume inl {
           g[(((g.s0.y.min_3*g.stride.1) + (((g.s0.x.x*4) + g.s0.x.v0) + g.min.0)) - ((g.min.1*g.stride.1) + g.min.0))] = inl[0]
          }
         }
        }
       }
      }
     }
    }
   }
  }
 }
}
```

**Comment by abadams** [comment]
I just hit a similar problem in the remove_max_on_fuse_factor branch. Bounds is just really bad at exploiting if conditions. We should try to not rely on that. I have an intrinsic promise_clamped in the branch add_image_checks_after_bounds_inference that might help. It just returns the arg, but declares that it's bounded for the purposes of bounds inference. I'll try it in remove_max_on_fuse_factor. If it works we might also want to use it in inside the compute_with predicates.

**Comment by vksnk** [comment]
> I just hit a similar problem in the remove_max_on_fuse_factor branch. Bounds is just really bad at exploiting if conditions. We should try to not rely on that. I have an intrinsic promise_clamped in the branch add_image_checks_after_bounds_inference that might help. It just returns the arg, but declares that it's bounded for the purposes of bounds inference. I'll try it in remove_max_on_fuse_factor. If it works we might also want to use it in inside the compute_with predicates.

That's a very cool idea and definitely would be a better solution. Hope it works out! 

I will add a test for my case in the meantime.

**Comment by abadams** [comment]
After some more investigation, I think the root of the problem might be that computation bounds inference walks inside let stmts, but doesn't walk inside ifs, so it makes an unnecessarily messy expression for the bounds depending on the nesting order of ifs and lets. It should really just walk inside both, and then the order won't matter as much, and it won't even try to understand an if (or a promise_clamped). Try cherry-picking 75dd135a983cdd4d5aa6560a0ad49698328038b0

**Comment by abadams** [comment]
Yeah if I just apply that change to master this case starts working cleanly, because all the bounds inference stuff happens inside the compute_with predicates.

**Comment by steven-johnson** [comment]
Is this still a desirable change, or has been obsoleted by the other one?

**Comment by vksnk** [comment]
No, there is a better fix for the original issue, so I'll close this one. We may want to update comments though, as some of them seem to be stale, I can do it separately.

**Pull request closed without merging**